File: app/src/main/java/com/example/themoviedb/popular_people_screen/popular_people_view/PopularPeopleActivity.java
package com.example.themoviedb.popular_people_screen.popular_people_view;

import android.support.annotation.NonNull;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.ProgressBar;

import com.example.themoviedb.R;
import com.example.themoviedb.others.Utilities;
import com.example.themoviedb.popular_people_screen.popular_people_model.PopularPeopleModel;
import com.example.themoviedb.popular_people_screen.popular_people_presenter.PopularPeoplePresenter;
import com.example.themoviedb.popular_people_screen.popular_people_model.PopularPeople;
import java.util.ArrayList;
import java.util.List;

public class PopularPeopleActivity extends AppCompatActivity implements android.widget.SearchView.OnQueryTextListener , PopularPeopleViewInterface {

    Boolean isScrolling = false ;
    int currentItems , totalItems , scrollingOutItems ;
    public int  page = 1 ;
    private RecyclerView recyclerView;
    private PopularPeopleAdapter popularPeopleAdapter;
    private LinearLayoutManager layoutManager ;
    SwipeRefreshLayout mSwipeRefreshLayout;
    private List<PopularPeople> popularPeopleList ;
    ProgressBar progressBar;
    android.widget.SearchView searchView;

    public Boolean isSearchAction= false;
    private String searchstr = "";
    private String pageStr="";
    MenuItem menuItem;

    PopularPeoplePresenter popularPeoplePresenter ;
    Utilities utilities;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_popular_people);

        recyclerView = (RecyclerView) findViewById(R.id.recyclerViewId);
        progressBar = findViewById(R.id.progress_bar_id);
        layoutManager = new LinearLayoutManager(this);
        popularPeopleList =new ArrayList<>();
        popularPeopleAdapter = new PopularPeopleAdapter(PopularPeopleActivity.this, popularPeopleList);
        recyclerView.setAdapter(popularPeopleAdapter);

        utilities = new Utilities();
        pageStr = String.valueOf(page);
        popularPeoplePresenter = new PopularPeoplePresenter(this , new PopularPeopleModel());

        popularPeoplePresenter.callFetchingData (utilities.popularPeopleURL+pageStr);

        mSwipeRefreshLayout = (SwipeRefreshLayout) findViewById(R.id.swipe_container);
        mSwipeRefreshLayout.setColorSchemeResources(R.color.colorAccent,
                android.R.color.holo_green_dark,
                android.R.color.holo_orange_dark,
                android.R.color.holo_blue_dark);

        mSwipeRefreshLayout.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {
            @Override
            public void onRefresh() {
                popularPeoplePresenter.callSwipeFun ();
              mSwipeRefreshLayout.setRefreshing(false);
            }
        });

        recyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrollStateChanged(@NonNull RecyclerView recyclerView, int newState) {
                super.onScrollStateChanged(recyclerView, newState);
                isScrolling = true ;
            }

            @Override
            public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy) {
                super.onScrolled(recyclerView, dx, dy);
                currentItems = layoutManager.getChildCount();
                scrollingOutItems = layoutManager.findFirstVisibleItemPosition();
                totalItems = layoutManager.getItemCount();
                if (isScrolling && (currentItems + scrollingOutItems == totalItems)) {
                    isScrolling = false;
                    progressBar.setVisibility(View.VISIBLE);
                    popularPeoplePresenter.callScrollingFun();
                }
            }
        });
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.my_menu, menu);
        menuItem = menu.findItem(R.id.search_btn);
        searchView = (android.widget.SearchView) menuItem.getActionView();
        searchView.setOnQueryTextListener(this);
        searchView.clearFocus();
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onQueryTextChange(String s) {
        if (searchView.getQuery().length() == 0) {
            isSearchAction = false;
            popularPeopleList.clear();
            searchView.clearFocus();
            popularPeoplePresenter.callFetchingData (utilities.popularPeopleURL+pageStr);
        }
        return true;
    }

    @Override
    public boolean onQueryTextSubmit(String s) {
        searchstr = s;
        if(!s.equals("")) {
            popularPeoplePresenter.clearList(popularPeopleList);
            isSearchAction = true;
            popularPeoplePresenter.callFetchingData (utilities.search_url + s);  ;
        }
        searchView.clearFocus();
        return true;
    }

    public void setPerson(PopularPeople person) {
        popularPeopleList.add(person);
    }

    public void settingAdapterInList() {
        popularPeopleAdapter.notifyDataSetChanged();
        recyclerView.setLayoutManager(layoutManager);
    }

    @Override
    public List<PopularPeople> getList() {
        return popularPeopleList ;
    }

    @Override
    public void notifyDataRemoved (int size) {
        popularPeopleAdapter.notifyItemRangeRemoved(0, size);

    }

    @Override
    public Boolean getSearchState() {
        return isSearchAction ;
    }
}

File: app/src/main/java/com/example/themoviedb/popular_people_screen/popular_people_view/PopularPeopleActivity.kt
package com.example.themoviedb.popular_people_screen.popular_people_view

import android.content.Intent
import android.support.v4.widget.SwipeRefreshLayout
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.ProgressBar

import com.example.themoviedb.R
import com.example.themoviedb.others.Utilities
import com.example.themoviedb.person_details_screen.person_details_view.PersonDetailsActivity
import com.example.themoviedb.popular_people_screen.popular_people_model.PopularPeopleModel
import com.example.themoviedb.popular_people_screen.popular_people_presenter.PopularPeoplePresenter
import com.example.themoviedb.popular_people_screen.popular_people_model.PopularPeople
import java.util.ArrayList

class PopularPeopleActivity : AppCompatActivity(), android.widget.SearchView.OnQueryTextListener, PopularPeopleViewInterface {

    internal var isScrolling: Boolean? = false
    internal var currentItems: Int = 0
    internal var totalItems: Int = 0
    internal var scrollingOutItems: Int = 0
    var page = 1
    private var recyclerView: RecyclerView? = null
    private var popularPeopleAdapter: PopularPeopleAdapter? = null
    private var layoutManager: LinearLayoutManager? = null
    internal lateinit var mSwipeRefreshLayout: SwipeRefreshLayout
    private var popularPeopleList: MutableList<PopularPeople>? = null
    internal lateinit var progressBar: ProgressBar
    internal lateinit var searchView: android.widget.SearchView

    var isSearchAction: Boolean? = false
    private var searchstr = ""
    private var pageStr = ""
    internal lateinit var menuItem: MenuItem

    internal lateinit var popularPeoplePresenter: PopularPeoplePresenter
    internal lateinit var utilities: Utilities

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_popular_people)

        recyclerView = findViewById<View>(R.id.recyclerViewId) as RecyclerView
        progressBar = findViewById(R.id.progress_bar_id)
        layoutManager = LinearLayoutManager(this)
        popularPeopleList = ArrayList()
        popularPeopleAdapter = PopularPeopleAdapter(this@PopularPeopleActivity, popularPeopleList!!)
        recyclerView!!.adapter = popularPeopleAdapter

        utilities = Utilities()
        pageStr = page.toString()
        popularPeoplePresenter = PopularPeoplePresenter(this, PopularPeopleModel())

        popularPeoplePresenter.callFetchingData()

        mSwipeRefreshLayout = findViewById<View>(R.id.swipe_container) as SwipeRefreshLayout
        mSwipeRefreshLayout.setColorSchemeResources(R.color.colorAccent,
                android.R.color.holo_green_dark,
                android.R.color.holo_orange_dark,
                android.R.color.holo_blue_dark)

        mSwipeRefreshLayout.setOnRefreshListener {
            popularPeoplePresenter.callSwipeFun()
            mSwipeRefreshLayout.isRefreshing = false
        }

        recyclerView!!.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrollStateChanged(recyclerView: RecyclerView, newState: Int) {
                super.onScrollStateChanged(recyclerView, newState)
                isScrolling = true
            }

            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                super.onScrolled(recyclerView, dx, dy)
                currentItems = layoutManager!!.childCount
                scrollingOutItems = layoutManager!!.findFirstVisibleItemPosition()
                totalItems = layoutManager!!.itemCount
                if (isScrolling!! && currentItems + scrollingOutItems == totalItems) {
                    isScrolling = false
                    progressBar.visibility = View.VISIBLE
                    popularPeoplePresenter.callScrollingFun()
                }
            }
        })
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.my_menu, menu)
        menuItem = menu.findItem(R.id.search_btn)
        searchView = menuItem.actionView as android.widget.SearchView
        searchView.setOnQueryTextListener(this)
        searchView.clearFocus()
        return super.onCreateOptionsMenu(menu)
    }

    override fun onQueryTextChange(s: String): Boolean {
        if (searchView.query.length == 0) {
            isSearchAction = false
            searchView.clearFocus()
            popularPeoplePresenter.callFetchingData()
        }
        return true
    }

    override fun onQueryTextSubmit(s: String): Boolean {
        searchstr = s
        if (s != "") {
            popularPeoplePresenter.clearList(popularPeopleList!!)
            isSearchAction = true
            popularPeoplePresenter.searchData(s)
        }
        searchView.clearFocus()
        return true
    }

    override fun setPerson(person: PopularPeople) {
        popularPeopleList!!.add(person)
    }

    override fun settingAdapterInList() {
        popularPeopleAdapter!!.notifyDataSetChanged()
        recyclerView!!.layoutManager = layoutManager
    }

    override fun getList(): List<PopularPeople>? {
        return popularPeopleList
    }

    override fun notifyDataRemoved(size: Int) {
        popularPeopleAdapter!!.notifyItemRangeRemoved(0, size)

    }

    override fun getSearchState(): Boolean? {
        return isSearchAction
    }
}